File: kotlin/src/commonMain/kotlin/com/xebia/functional/textsplitters/CharacterTextSplitter.kt
package com.xebia.functional.textsplitters

import com.xebia.functional.Document

suspend fun CharacterTextSplitter(
    separator: String
): BaseTextSplitter = object : BaseTextSplitter {

    override suspend fun splitText(text: String): List<String> =
        text.split(separator)

    override suspend fun splitDocuments(documents: List<Document>): List<Document> =
        documents.flatMap { doc -> doc.content.split(separator) }.map(::Document)

    override suspend fun splitTextInDocuments(text: String): List<Document> =
        text.split(separator).map(::Document)
}

File: scala/src/main/scala/com/xebia/functional/textsplitters/CharacterTextSplitter.scala
package com.xebia.functional.scala.textsplitters

import cats.Applicative
import cats.ApplicativeThrow
import cats.syntax.all.*

import com.xebia.functional.scala.domain.Document

class CharacterTextSplitter[F[_]: ApplicativeThrow](separator: String) extends BaseTextSplitter[F]:
  def splitText(text: String): F[List[String]] =
    ApplicativeThrow[F]
      .catchNonFatal(
        text.split(separator).toList
      ).adaptErr(toCharacterSplitterError)

  def splitDocuments(documents: List[Document]): F[List[Document]] =
    ApplicativeThrow[F]
      .catchNonFatal(
        documents.flatMap(doc => doc.content.split(separator).toList.map(Document.apply))
      ).adaptErr(toCharacterSplitterError)

  def splitTextInDocuments(text: String): F[List[Document]] =
    ApplicativeThrow[F]
      .catchNonFatal(
        text.split(separator).toList.map(Document.apply)
      ).adaptErr(toCharacterSplitterError)

  private def toCharacterSplitterError: PartialFunction[Throwable, Throwable] = { case e: Throwable =>
    SplitterError(Option(e.getMessage).getOrElse("<null>"))
  }
